# Segmenting and Clustering Neighborhoods in Toronto

# Assignment 1. Download Dataset in Wikipedia page

In [94]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [95]:
r=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

#### Parsing the web html file with BeautifulSoup package:

In [96]:
page=bsoup(r.text,"html.parser")
page

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgMonthNamesShort":["","Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"wgRequestId":"Xg1lcwpAICgAALHCp4UAAADQ","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":933624196,"wgRevisionId":933624196,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontari

'The htlm 'page' indicates that there is a 'table class' - let's find that table

In [97]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Harbourfront</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Lawrence_Heights" title="Lawrence Heights">Lawrence Heights</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North_

In [98]:
results=rtable.find_all('tr')
nrows=len(results)
nrows

288

Such table named 'results' contains 288 rows. The first tow is the table header

In [99]:
results[0:5]

[<tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighborhood
 </th></tr>, <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>, <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>]

##### The Header of the DataFrame is the first row of data:

In [100]:
header=results[0].text.split()
header

['Postcode', 'Borough', 'Neighborhood']

Let's check some rows in order to prepare to build the loop that will extract all cells into a DataFrame.
For example, let's examine row 129

In [101]:
results[129].text

'\nM4M\nEast Toronto\nStudio District\n'

In [102]:
results[129].text.split('\n')

['', 'M4M', 'East Toronto', 'Studio District', '']

In [103]:
Postcode=results[129].text.split('\n')[1]
Postcode

'M4M'

In [104]:
Borough=results[129].text.split('\n')[2]
Borough

'East Toronto'

In [105]:
Neighborhood=results[129].text.split('\n')[3]
Neighborhood

'Studio District'

###### Iteration loop to extract all cells into a dataframe df:

In [106]:
records =[]
n=1
while n < nrows :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[2]
    Neighborhood=results[n].text.split('\n')[3]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighborhood'])
df.head(5)

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [107]:
df.shape

(287, 3)

In [108]:
# How many rows have Borough equal to 'Not assigned'?
df[df['Borough']=='Not assigned'].count()

PostalCode      77
Borough         77
Neighborhood    77
dtype: int64

Initial Dataframe has 287 cells or rows of data, as expected.

Now, let's remove all rows with Borough = 'Not assigned' strings

In [109]:
df_Tor_Neighb=df[~df.Borough.str.contains("Not assigned")]
df_Tor_Neighb=df_Tor_Neighb.reset_index(drop=True)

In [110]:
df_Tor_Neighb.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

In [111]:
df_Tor_Neighb.shape

(210, 3)

In [112]:
df_Tor_Neighb.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

In [113]:
df_Tor_Neighb.tail()

PostalCode    Borough              Neighborhood
205        M8Z  Etobicoke  Kingsway Park South West
206        M8Z  Etobicoke                 Mimico NW
207        M8Z  Etobicoke        The Queensway West
208        M8Z  Etobicoke     Royal York South West
209        M8Z  Etobicoke            South of Bloor

The dataframe now has 210 rows (0-209) after droppong all Borough with 'Not assigned', which is correct ( 287-77)¶

In [114]:
df_Tor_Neighb.describe()

PostalCode    Borough    Neighborhood
count         210        210             210
unique        103         11             208
top           M9V  Etobicoke  St. James Town
freq            8         44               2

In [115]:
postalcodes = df_Tor_Neighb['PostalCode'].nunique()
boroughs = df_Tor_Neighb['Borough'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))

Unique Postalcodes : 103
Unique Boroughs  : 11


More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma:

In [117]:
df_Tor_Neigh= df_Tor_Neighb.groupby(['PostalCode', 'Borough'])['Neighborhood'].agg(lambda col: ', '.join(col.astype('str'))).reset_index()

In [118]:
df_Tor_Neigh.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Rouge, Malvern
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

### Dataframe has 103 rows and 3 columns:

In [119]:
df_Tor_Neigh.shape

(103, 3)

## End of Assignment 1¶

__________________________________________________________________________________

# Assignment 2. Get the latitude and the longitude coordinates of each neighborhood

NOTE: I tried the geocoder but it did not work reliably, therefore I decided to use the cvs file with postalcodes provided

#### Read the cvs file and convert it to a datafram:

In [120]:
df_pcodes=pd.read_csv('D:\Geospatial_Coordinates.csv')
df_pcodes.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Need to rename the column "Postal Code" to "PostalCode" in order to do a proper merger

In [121]:
df_pcodes.columns = ['PostalCode', 'Latitude', 'Longitude']
df_pcodes.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [122]:
df_Coor_Toronto=df_Tor_Neigh.merge(df_pcodes) #combine two tables into one

In [123]:
df_Coor_Toronto.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [124]:
df_Coor_Toronto.shape

(103, 5)

### Table dataframe below contains 103 rows by the 5 columns requested.

## End of Assignment 2.

________________________________________________________________________

# Assignment 3. Explore and cluster the neighborhoods in Toronto

In [125]:
!conda install -c conda-forge folium=0.5.0 --yes

import folium

print('Folium installed and imported!')

Solving environment: ...working... done

# All requested packages already installed.

Folium installed and imported!


##### Visualizat Central, West, East and Downtown Toronto the neighborhoods in it:

In [126]:
Toronto_data= df_Coor_Toronto[df_Coor_Toronto['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)

In [127]:
Toronto_data.head()

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [128]:
Toronto_data.shape

(39, 5)

In [129]:
Toronto_map = folium.Map(location=[43.6520,-79.3952], zoom_start=12)

for lat, lng, label in zip(Toronto_data['Latitude'], Toronto_data['Latitude'], Toronto_data['Borough']):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5, # define how big you want the circle markers to be
        color='yellow',
        fill=True,
        popup=label,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(Toronto_map)

# show map
Toronto_map

In [37]:
!conda install -c conda-forge geopy --yes 

Solving environment: ...working... done

# All requested packages already installed.



In [130]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

Define Foursquare Credentials and Version

In [132]:
import json # library to handle JSON files

In [135]:
Toronto_data.loc[0, 'Neighborhood']

'The Beaches'

In [136]:
neighborhood_latitude = Toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = Toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = Toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


Get the top 100 venues that are in The Beaches within a radius of 500 meters.

In [137]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius
 # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

In [138]:
results = requests.get(url).json()

In [139]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [140]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat  \
0                   Glen Manor Ravine                 Trail  43.676821   
1  The Big Carrot Natural Food Market     Health Food Store  43.678879   
2                 Grover Pub and Grub                   Pub  43.679181   
3                 Glen Stewart Ravine  Other Great Outdoors  43.676300   
4                       Upper Beaches          Neighborhood  43.680563   

         lng  
0 -79.293942  
1 -79.297734  
2 -79.297215  
3 -79.294784  
4 -79.292869

In [141]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

6 venues were returned by Foursquare.


#### Explore Neighborhoods in Toronto

In [142]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [143]:
Toronto_venues = getNearbyVenues(names=Toronto_data['Neighborhood'],
                                   latitudes=Toronto_data['Latitude'],
                                   longitudes=Toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The Junction Sout

Check the size of the resulting dataframe:

In [144]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1721, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                 Glen Stewart Ravine       43.676300       -79.294784   
4                       Upper Beaches       43.680563       -79.292869   

         Venue Category  
0                 Trail  
1     Health Food Store  
2                   Pub  
3  Other Great Outdoors  
4          Neighborhood

In [145]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [146]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            57   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      18   
CN Tower, Bathurst Quay, Island airport, Harbou...                     12   
Cabbagetown, St. James Town                                            43   
Central Bay Street                                                     82   
Chinatown, Grange Park, Kensington Market                              92   
Christie                                                               17   
Church and Wellesley                                                   85   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             30   
Davisville North                                                        7   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           18   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         36   
Harbourfront                                                           48   
Harbourfront East, Toronto Islands, Union Station                     100   
High Park, The Junction South                                          25   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               55   
Moore Park, Summerhill East                                             3   
North Toronto West                                                     22   
Parkdale, Roncesvalles                                                 13   
Queen's Park                                                           41   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     40   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        97   
Studio District                                                        42   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             6   
The Beaches West, India Bazaar                                         19   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             57   
Brockton, Exhibition Place, Parkdale Village                            22   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower, Bathurst Quay, Island airport, Harbou...                      12   
Cabbagetown, St. James Town                                             43   
Central Bay Street                                                      82   
Chinatown, Grange Park, Kensington Market                               92   
Christie                                                                17   
Church and Wellesley                

Find out how many unique categories can be curated from all the returned venues:

In [147]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 234 uniques categories.


### Analyze Each Neighborhood

In [148]:
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] =Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Lounge  Airport Service  \
0            0        0                   0               0                0   
1            0        0                   0               0                0   
2            0        0                   0               0                0   
3            0        0                   0               0                0   
4            0        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Candy Store  Caribbean Restaurant  \
0                          0             0            0                     0   
1                          0             0            0                     0   
2                          0             0            0                     0   
3                          0             0            0                     0   
4                   

In [149]:
Toronto_onehot.shape

(1721, 234)

In [150]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio   Airport  \
0                            Adelaide, King, Richmond     0.000000  0.000000   
1                                         Berczy Park     0.000000  0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.045455  0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.055556  0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000  0.083333   
5                         Cabbagetown, St. James Town     0.000000  0.000000   
6                                  Central Bay Street     0.012195  0.000000   
7           Chinatown, Grange Park, Kensington Market     0.000000  0.000000   
8                                            Christie     0.000000  0.000000   
9                                Church and Wellesley     0.023529  0.000000   
10                     Commerce Court, Victoria Hotel     0.000000  0.000000   
11                                         Davisville     0.000000  0.000000   
12                                   Davisville North     0.000000  0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000  0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000  0.000000   
15                       Dovercourt Village, Dufferin     0.000000  0.000000   
16             First Canadian Place, Underground city     0.000000  0.000000   
17                Forest Hill North, Forest Hill West     0.000000  0.000000   
18                     Harbord, University of Toronto     0.000000  0.000000   
19                                       Harbourfront     0.020833  0.000000   
20  Harbourfront East, Toronto Islands, Union Station     0.000000  0.000000   
21                      High Park, The Junction South     0.000000  0.000000   
22                                      Lawrence Park     0.000000  0.000000   
23                           Little Portugal, Trinity     0.018182  0.000000   
24                        Moore Park, Summerhill East     0.000000  0.000000   
25                                 North Toronto West     0.045455  0.000000   
26                             Parkdale, Roncesvalles     0.000000  0.000000   
27                                       Queen's Park     0.024390  0.000000   
28                                           Rosedale     0.000000  0.000000   
29                                           Roselawn     0.000000  0.000000   
30                                 Runnymede, Swansea     0.000000  0.000000   
31                           Ryerson, Garden District     0.000000  0.000000   
32                                     St. James Town     0.000000  0.000000   
33                    Stn A PO Boxes 25 The Esplanade     0.000000  0.000000   
34                                    Studio District     0.023810  0.000000   
35                The Annex, North Midtown, Yorkville     0.000000  0.000000   
36                                        The Beaches     0.000000  0.000000   
37                     The Beaches West, India Bazaar     0.000000  0.000000   
38                       The Danforth West, Riverdale     0.023810  0.000000   

    Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0             0.000000        0.000000         0.000000          0.000000   
1             0.000000        0.000000         0.000000          0.000000   
2             0.000000        0.000000         0.000000          0.000000   
3             0.000000        0.000000         0.000000          0.000000   
4             0.083333        0.166667         0.166667          0.083333   
5             0.000000        0.000000         0.000000          0.000000   
6             0.000000        0.000000         0.000000          0.000000   
7             0.000000        0.000000         0.000000          0.000000   
8             0.000000        0.000000         0.000000          0.000000   
9             0.000000        0.000000         0.000000          0.000

In [151]:
Toronto_grouped.shape

(39, 234)

In [152]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
         venue  freq
0  Coffee Shop  0.08
1         Café  0.05
2   Steakhouse  0.04
3          Bar  0.04
4   Restaurant  0.03


----Berczy Park----
            venue  freq
0     Coffee Shop  0.09
1    Cocktail Bar  0.05
2  Farmers Market  0.04
3      Steakhouse  0.04
4        Beer Bar  0.04


----Brockton, Exhibition Place, Parkdale Village----
            venue  freq
0            Café  0.14
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3     Yoga Studio  0.05
4    Intersection  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2                 Spa  0.06
3       Garden Center  0.06
4              Garden  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
                venue  freq
0      Airport Lounge  0.17
1     Airport Service  0.17
2       Boat or Ferry  0.08
3  Airport Food Cou

In [153]:
# function to sort the venues in descending order
def return_most_common_venues(row, num_top_venues): 
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighborhood

In [167]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse                   Bar   
1          Cocktail Bar                Bakery           Cheese Shop   
2           Coffee Shop        Breakfast Spot           Yoga Studio   
3         Auto Workshop            Skate Park            Smoke Shop   
4       Airport Service              Boutique      Airport Terminal   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant      Asian Restaurant      Sushi Restaurant   
1                  Café        Farmers Market              Beer Bar   
2               Stadium         Burrito Place            Restaurant   
3         Burrito Place                   Spa        Farmers Market   
4         Boat or Ferry       Harbor / Marina      Sculpture Garden   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant          Burger Joint                    Gym  
1            Steakhouse    Seafood Restaurant               Creperie  
2          Climbing Gym             Pet Store                 Bakery  
3  Fast Food Restaurant               Butcher             Restaurant  
4    Airport Food Court               Airport    Rental Car Location

### Cluster Neighborhoods

Run k-means to cluster the neighborhood into 4 clusters.

In [178]:
kclusters = 6

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([5, 5, 1, 1, 1, 1, 5, 1, 1, 5])

In [169]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               5                   Pub  Other Great Outdoors   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               1        Sandwich Place                  Park   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               1                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop     Health Food Store                 Trail   
1    Italian Restaurant        Ice Cream Shop            Restaurant   
2                   Pub          Burger Joint         Burrito Place   
3   American Restaurant                Bakery    Italian Restaurant   
4              Bus Line    Dim Sum Restaurant          Concert Hall   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Women's Store                 Diner        Discount Store   
1  Furniture / Home Store           Yoga Studio     Indian Restaurant   
2            Liquor Store    Italian Restaurant  Fast Food Restaurant   
3               Gastropub  Gym / Fitness Center           Fish Market   
4       Convenience Store           Event Space  Ethiopian Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run       Doner Restaurant  
1                   Pub                  Diner  
2     Fish & Chips Shop             Steakhouse  
3           Pizza Place              Pet Store  
4   Empanada Restaurant      Electronics Store

In [170]:
map_clusters = folium.Map(location=[43.6520,-79.3952], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

###  Examine Clusters
Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. 

In [171]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto               0                   Gym   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8            Playground                 Trail          Dessert Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8           Event Space  Ethiopian Restaurant   Empanada Restaurant   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
8     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

In [172]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       East Toronto               1      Greek Restaurant   
2       East Toronto               1        Sandwich Place   
3       East Toronto               1                  Café   
4    Central Toronto               1                  Park   
7    Central Toronto               1        Sandwich Place   
11  Downtown Toronto               1           Coffee Shop   
24   Central Toronto               1        Sandwich Place   
25  Downtown Toronto               1                  Café   
26  Downtown Toronto               1                  Café   
27  Downtown Toronto               1        Airport Lounge   
30  Downtown Toronto               1         Grocery Store   
31      West Toronto               1           Supermarket   
32      West Toronto               1                   Bar   
33      West Toronto               1                  Café   
34      West Toronto               1         Grocery Store   
35      West Toronto               1             Gift Shop   
36      West Toronto               1           Coffee Shop   
38      East Toronto               1    Light Rail Station   

    2nd Most Common Venue          3rd Most Common Venue  \
1             Coffee Shop             Italian Restaurant   
2                    Park                            Pub   
3             Coffee Shop            American Restaurant   
4             Swim School                       Bus Line   
7            Dessert Shop                    Pizza Place   
11                   Café                            Pub   
24                   Café                    Coffee Shop   
25    Japanese Restaurant                            Bar   
26  Vietnamese Restaurant  Vegetarian / Vegan Restaurant   
27        Airport Service                       Boutique   
30                   Café                           Park   
31               Pharmacy                         Bakery   
32       Asian Restaurant                     Restaurant   
33            Coffee Shop                 Breakfast Spot   
34                   Café             Mexican Restaurant   
35              Bookstore                  Movie Theater   
36                   Café             Italian Restaurant   
38          Auto Workshop                     Skate Park   

          4th Most Common Venue      5th Most Common Venue  \
1                Ice Cream Shop                 Restaurant   
2                  Burger Joint              Burrito Place   
3                        Bakery         Italian Restaurant   
4            Dim Sum Restaurant               Concert Hall   
7            Italian Restaurant                        Gym   
11           Italian Restaurant                Pizza Place   
24                  Pizza Place  Middle Eastern Restaurant   
25                    Bookstore         Italian Restaurant   
26           Chinese Restaurant                Coffee Shop   
27             Airport Terminal              Boat or Ferry   
30                  Coffee Shop         Italian Restaurant   
31         Gym / Fitness Center     Furniture / Home Store   
32                  Coffee Shop                       Café   
33                  Yoga Studio                    Stadium   
34                          Bar            Thai Restaurant   
35  Eastern European Restaurant         Italian Restaurant   
36            Electronics Store           Sushi Restaurant   
38                   Smoke Shop              Burrito Place   

     6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
1   Furniture / Home Store           Yoga Studio      Indian Restaurant   
2             Liquor Store    Italian Restaurant   Fast Food Restaurant   
3                Gastropub  Gym / Fitness Center            Fish Market   
4        Convenience Store           Event Space   Ethiopian Restaurant   
7              Coffee Shop                  Café       Sushi Restaurant   
11                  Bakery            Restaurant    Japanese Res

In [173]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10            Playground                 Trail      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
10  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

In [174]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22         Women's Store          Dessert Shop           Event Space   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

In [175]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23  Central Toronto               4         Jewelry Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23                 Trail              Bus Line      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23         Women's Store                 Diner           Event Space   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23  Ethiopian Restaurant   Empanada Restaurant      Electronics Store

In [176]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 5, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               5                   Pub   
5    Central Toronto               5                   Gym   
6    Central Toronto               5        Clothing Store   
9    Central Toronto               5           Coffee Shop   
12  Downtown Toronto               5           Coffee Shop   
13  Downtown Toronto               5           Coffee Shop   
14  Downtown Toronto               5           Coffee Shop   
15  Downtown Toronto               5                  Café   
16  Downtown Toronto               5           Coffee Shop   
17  Downtown Toronto               5           Coffee Shop   
18  Downtown Toronto               5           Coffee Shop   
19  Downtown Toronto               5           Coffee Shop   
20  Downtown Toronto               5           Coffee Shop   
21  Downtown Toronto               5           Coffee Shop   
28  Downtown Toronto               5           Coffee Shop   
29  Downtown Toronto               5           Coffee Shop   
37  Downtown Toronto               5           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0   Other Great Outdoors           Coffee Shop     Health Food Store   
5         Clothing Store        Breakfast Spot     Food & Drink Shop   
6            Coffee Shop   Sporting Goods Shop  Gym / Fitness Center   
9                    Pub           Pizza Place    Light Rail Station   
12      Sushi Restaurant               Gay Bar   Japanese Restaurant   
13                   Pub                Bakery                  Park   
14        Clothing Store                  Café        Cosmetics Shop   
15           Coffee Shop            Restaurant                 Hotel   
16          Cocktail Bar                Bakery           Cheese Shop   
17    Italian Restaurant                  Café        Ice Cream Shop   
18                  Café            Steakhouse                   Bar   
19              Aquarium                 Hotel                  Café   
20                  Café                 Hotel            Restaurant   
21                  Café                 Hotel            Restaurant   
28                  Café            Restaurant          Cocktail Bar   
29                  Café            Steakhouse            Restaurant   
37                  Park                   Gym     College Cafeteria   

   5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0                  Trail              Women's Store                 Diner   
5         Sandwich Place                      Hotel                  Park   
6     Mexican Restaurant         Salon / Barbershop            Restaurant   
9           Liquor Store                 Sports Bar            Restaurant   
12            Restaurant                Yoga Studio                  Café   
13            Restaurant         Mexican Restaurant        Breakfast Spot   
14                Bakery  Middle Eastern Restaurant            Restaurant   
15        Clothing Store             Breakfast Spot          Cocktail Bar   
16                  Café             Farmers Market              Beer Bar   
17        Sandwich Place               Burger Joint  Gym / Fitness Center   
18       Thai Restaurant           Asian Restaurant      Sushi Restaurant   
19    Italian Restaurant                    Brewery        Scenic Lookout   
20            Steakhouse                  Gastropub    Seafood Restaurant   
21   American Restaurant              Deli / Bodega             Gastropub   
28   Japanese Restaurant                      Hotel    Seafood Restaurant   
29                   Gym                      Hotel   Japanese Restaurant   
37         Smoothie Shop             Sandwich Place          Burger Joint   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Discount Store               Dog Run       Doner Restaurant  
5               Creperie       Coworking Space             Comic Sho

### Thank you for checking out this work!